### DISCUSSION TITLE
Propel2 logging

### DISCUSSION THREAD

**Discussion by murraycollingwood** [propelorm/Propel2#1871] [Q&A] [answered]
Sorry @mringler but you keep creating more questions for me.

Setting up Propel2 logging, now I did get this working and it's nice, but there are some probs with Intelephense understanding your coding.

There are two entries, this is the first.  Intelephense is having trouble locating the function setLogger(), and so was I:
<img width="506" alt="image" src="https://user-images.githubusercontent.com/5810216/164009433-7791eef3-ee2a-482b-8841-1f47fb89408f.png">

Propel::getServiceContainer returns a "ServiceContainerInterface".  Well, that seems odd, returning an interface.  Shouldn't it return a substantiating class rather than an interface?  ie Should it return the StandardServiceContainer.  ServiceContainerInterface doesn't implement "setLogger()" and hence why Intelephense doesn't understand the function.

The second entry:
<img width="439" alt="image" src="https://user-images.githubusercontent.com/5810216/164010402-04b457ad-84a8-4877-8a4a-e17caa22c586.png">

This a similar problem.  Propel::getWriteConnection() returns a ConnectionInterface.  ConnectionInterface does not implement "useDebug()" and hence why Intelephense doesn't understand the function.

The code does actually work, but only because those function don't return an interface so much as a class that implements the interface.  I guess that works, for now, but how many versions of PHP until support for this functionality is deprecated and removed.

**Comment by mringler** [comment]
Yes, I agree completely, this hidden functionality is very annoying. I think the idea behind the structure was to allow users to replace these classes with their own implementations, a feature that probably next to no one uses, at the price of obscuring access to very common and important functionality. Then again, the problem ultimately comes down to the limited type system in PHP.
Not sure what a clean fix for this would be.

Speaking of hidden functionality, I just found out that there is an easier way to set the logger:
```php
$loggerConfiguration = [
    'type' => 'stream',
    'path' => 'path/to/file'
    'level' => Logger::INFO, // Defaults to Logger::CRITICAL,
    'bubble' => true, // Defaults to true
];
Propel::getServiceContainer()->setLoggerConfiguration('defaultLogger', $loggerConfiguration);
```

**Comment by mringler** [comment, marked as answer]
I have created two small PRs to address the issues with type checking (and IDE support):
https://github.com/propelorm/Propel2/pull/1872
https://github.com/propelorm/Propel2/pull/1873
Let's see how that goes. If you have time for a review, it is surely appreciated.

**Reply by murraycollingwood** [reply]
Thanks Moritz
